async_batcher/batcher.py

```diff
@@ -54,86 +48,66 @@ async def process_batch(self, batch: list[T]) -> list[S] | None:
-        result = await asyncio.get_event_loop().create_task(self._process_single(item=item))
-        if isinstance(result, Exception):
-            raise result
-        return result
-
-    async def _arun(self):
+        if self._current_task is None:
+            self._current_task = asyncio.get_running_loop().create_task(self.batch_run())
+        logging.debug(item)
+        future = asyncio.get_running_loop().create_future()
+        await self._queue.put(self.QueueItem(item, future))
+        await future
+        return future.result()
+
+    async def batch_run(self):
         """Run the batcher asynchronously."""
         self._is_running = True
-        while not self._should_stop or (not self._force_stop and len(self._buffer) > 0):
-            ids = []
-            batch = []
-            while self._buffer and (len(batch) < self.batch_size or self.batch_size == -1):
-                query_id, item = self._buffer.popleft()
-                batch.append(item)
-                ids.append(query_id)
-            if batch:
-                started_at = asyncio.get_event_loop().time()
+        while not self._should_stop or (not self._force_stop and self._queue.qsize() > 0):
+            try:
+                batch = [await asyncio.wait_for(self._queue.get(), timeout=1.0)]
+            except asyncio.TimeoutError:
+                continue
+            started_at = asyncio.get_running_loop().time()
+            while 1:
                 try:
-                    results = await asyncio.get_event_loop().create_task(self.process_batch(batch=batch))
-                    if results is None:
-                        results = [None] * len(batch)
-                    if len(results) != len(batch):
-                        raise ValueError(f"Expected to get {len(batch)} results, but got {len(results)}.")
-                except Exception as e:
-                    self.logger.error("Error processing batch", exc_info=True)
-                    results = [e] * len(batch)
-                for query_id, result in zip(ids, results):
-                    self._results[query_id] = result
-                elapsed_time = asyncio.get_event_loop().time() - started_at
-                self.logger.debug(
-                    f"Processed batch {self._current_batch} of {len(batch)} elements"
-                    f" in {elapsed_time} seconds."
-                )
-                self._current_batch += 1
-                await asyncio.sleep(max(self.buffering_time - elapsed_time, 0))
+                    max_wait = self.max_queue_time - (asyncio.get_running_loop().time() - started_at)
+                    if max_wait > 0:
+                        item = await asyncio.wait_for(self._queue.get(), timeout=self.max_queue_time)
+                    else:
+                        item = self._queue.get_nowait()
+                    batch.append(item)
+                    if self.max_batch_size is not None and 1 < self.max_batch_size <= len(batch):
```

Comment:
If max_batch_size is 1 (which would be strange), this check will consider it as infinite size. We can:
- change this to `0 < self.max_batch_size <= len(batch)` and move the check to after the try
- or add a check in the `__init__` to not accept 0 and 1 for `max_batch_size`.

I can do that in a separate PR